fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/IcebergEqualityDeleteRewriteRule.java

```diff
@@ -0,0 +1,370 @@
+    }
+
+    @Override
+    public boolean check(final OptExpression input, OptimizerContext context) {
+        Operator op = input.getOp();
+        if (!(op instanceof LogicalIcebergScanOperator)) {
+            return false;
+        }
+        LogicalIcebergScanOperator scanOperator = op.cast();
+        if (scanOperator.isFromEqDeleteRewriteRule()) {
+            return false;
+        }
+
+        Optional<Long> snapshotId = scanOperator.getTableVersionRange().end();
+        if (snapshotId.isEmpty()) {
+            return false;
+        }
+        IcebergTable icebergTable = (IcebergTable) scanOperator.getTable();
+        if (!icebergTable.isV2Format()) {
+            return false;
+        }
+
+        Table table = icebergTable.getNativeTable();
+        Snapshot snapshot = table.snapshot(snapshotId.get());
+        if (snapshot == null) {
+            return false;
+        }
+
+        boolean mayHaveEqualityDeletes = IcebergApiConverter.mayHaveEqualityDeletes(snapshot);
+        if (!mayHaveEqualityDeletes) {
+            return false;
+        }
+
+        // Use the same PredicateSearchKey to get iceberg scan tasks in the query level cache.
+        GetRemoteFilesParams params =
+                GetRemoteFilesParams.newBuilder().setTableVersionRange(scanOperator.getTableVersionRange())
+                        .setPredicate(scanOperator.getPredicate()).build();
+        List<RemoteFileInfo> splits = GlobalStateMgr.getCurrentState().getMetadataMgr().getRemoteFiles(icebergTable, params);
+        if (splits.isEmpty()) {
+            return false;
+        }
+
+        IcebergRemoteFileDesc remoteFileDesc = (IcebergRemoteFileDesc) splits.get(0).getFiles().get(0);
+        if (remoteFileDesc == null) {
+            return false;
+        }
+
+        for (FileScanTask fileScanTask : remoteFileDesc.getIcebergScanTasks()) {
+            for (DeleteFile deleteFile : fileScanTask.deletes()) {
+                if (deleteFile.content() == FileContent.EQUALITY_DELETES) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        LogicalIcebergScanOperator scanOperator = input.getOp().cast();
+        IcebergTable icebergTable = (IcebergTable) scanOperator.getTable();
+        Table table = icebergTable.getNativeTable();
+        Set<DeleteSchema> deleteSchemas = collectDeleteSchemas(scanOperator, icebergTable);
+        if (deleteSchemas.isEmpty()) {
+            scanOperator.setFromEqDeleteRewriteRule(true);
+            return Collections.singletonList(input);
+        }
+
+        long limit = scanOperator.getLimit();
+        ColumnRefFactory columnRefFactory = context.getColumnRefFactory();
+        boolean hasPartitionEvolution = deleteSchemas.stream().map(x -> x.specId).distinct().count() > 1;
```

Comment:
If the timeline of operation is as follows:
T1: insert data
T2: partition evolution
T3: delete data

the distinct spec id of the delete schema is 1 but it does partition evolution

Reply:
this only represents whether this query need to add spec_id to extended columns.

Reply:
So you mean the eq delete files generated in T3 can delete data in T1?

Reply:
no. the eq delete files generated in T3 won't be matched by any data files.